src/EditorFeatures/Core/DocumentationComments/DocumentationCommentSuggestion.cs

```diff
@@ -65,34 +66,122 @@ public override Task OnProposalUpdatedAsync(SuggestionSessionBase session, Propo
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// This is where we actually try to display the grey-text from the proposal
+        /// we created.
+        /// </summary>
+        public async Task TryDisplayDocumentationSuggestionAsync(ProposalBase proposal, CancellationToken cancellationToken)
         {
             try
             {
-                await providerInstance.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                await session.DisplayProposalAsync(Proposal, cancellationToken).ConfigureAwait(false);
-                return true;
+                await RunWithEnqueueActionAsync<bool>(
+                    "DisplayProposal",
+                    async () =>
+                    {
+                        await _suggestionSession!.DisplayProposalAsync(proposal, cancellationToken).ConfigureAwait(false);
+                        return true;
+                    },
+                    cancellationToken).ConfigureAwait(false);
+
+                Logger.Log(FunctionId.Copilot_Generate_Documentation_Displayed, logLevel: LogLevel.Information);
             }
             catch (OperationCanceledException)
             {
                 Logger.Log(FunctionId.Copilot_Generate_Documentation_Canceled, logLevel: LogLevel.Information);
             }
+        }
+
+        /// <summary>
+        /// Dismisses the session if the proposal we generated was invalid.
+        /// Needs to dispose of the IntelliCodeCompletionsDisposable so we no longer have exclusive right to
+        /// display any grey text.
+        /// </summary>
+        private async Task DismissSuggestionSessionAsync(CancellationToken cancellationToken)
+        {
+            await DisposeIntelliCodeCompletionsDisposableAsync().ConfigureAwait(false);
+            await RunWithEnqueueActionAsync<bool>(
+                "DismissSuggestionSession",
+                async () =>
+                {
+                    await _suggestionSession!.DismissAsync(ReasonForDismiss.DismissedDueToInvalidProposal, cancellationToken).ConfigureAwait(false);
+                    return true;
+                },
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// In general, calls to a SuggestionManager or SuggestionSession need to be wrapped in an EnqueueAction.
+        /// This is the pattern recommended by VS Platform to avoid races.
+        /// Pattern from platform shown here:
+        /// https://devdiv.visualstudio.com/DevDiv/_git/IntelliCode-VS?path=/src/VSIX/IntelliCode.VSIX/SuggestionService/AmbientAI/SuggestionProviderForAmbientAI.cs
+        /// </summary>
+        private async Task<T> RunWithEnqueueActionAsync<T>(string description, Func<Task<T>> action, CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(SuggestionManager);
+
+            var taskCompletionSource = new TaskCompletionSource<T>();
```

Comment:
Given the (slightly odd) design that this queue doesn't provide a way to track work, and you're hacking one in, you're potentially opening it up to a problem. When `action()` completes, continuations that your returned Task adds on (by awaiting the result of this method) may run inline with the action itself, holding up the queue even longer before it can work on the next thing.
You may want to create this TCS with `TaskCreationOptions.RunContinuationsAsynchronously` to prevent accidentally inlining arbitrary work into the queue.